api/sso/routes.py

```diff
@@ -87,10 +96,25 @@ def get_token(self):
             roles=session["user"].get("roles"),
         )
 
+        redirect_url = (
+            Config.ASSESSMENT_POST_LOGIN_URL
+        )  # TODO: Remove defaulting to Assessment, instead use return_app
+        if return_app := session.get("return_app"):
+            if safe_app := Config.SAFE_RETURN_APPS.get(return_app):
+                redirect_url = safe_app.login_url
+                current_app.logger.info(
+                    f"Returning to {return_app} @ {redirect_url}"
+                )
+            else:
+                current_app.logger.warning(
+                    f"{return_app} not listed as a safe app."
+                )
+                abort(400, "Unknown return app.")
+
         # Create session token, set cookie and redirect
         return AuthSessionView.create_session_and_redirect(
             account=updated_account,
-            redirect_url=Config.ASSESSMENT_POST_LOGIN_URL,
+            redirect_url=redirect_url,
             is_via_magic_link=False,
             timeout_seconds=Config.FSD_ASSESSMENT_SESSION_TIMEOUT_SECONDS,
         )
```

Comment:
Don't we need `return_app` here? As if we use the `SAFE_RETURN_APPS` of `post-award-frontend` we'll go to:

https://github.com/communitiesuk/funding-service-design-authenticator/blob/eb62551669cc80a0cae6672c37111c7eea905f7c/frontend/sso/routes.py#L16-L27

Unless it's okay to have `?return_app=None` on the back of that route?

```python3
>>> from urllib.parse import urlencode
>>> return_app = None
>>> f"?{urlencode({'return_app': return_app})}"
'?return_app=None'
```

Reply:
Good spot - `?return_app=None` would currently be handled by throwing a 400 error so this needs to change.

Reply:
It feels a bit wrong that we have to be careful of `?return_app=None` and I also don't like having to check if `return_app` exists all the time, think it will be much more succinct and clear once we force all apps to use `return_app` and then we can comfortably return 400 errors when there is `return_app=None`

Reply:
Yeah sounds good